rust_snuba/rust_arroyo/src/processing/strategies/reduce.rs

```diff
@@ -0,0 +1,195 @@
+use crate::processing::strategies::{CommitRequest, MessageRejected, ProcessingStrategy};
+use crate::types::{AnyMessage, InnerMessage, Message, Partition};
+use std::collections::BTreeMap;
+use std::time::{Duration, SystemTime};
+use std::mem;
+
+struct BatchState<T, TResult> {
+    value: Option<TResult>,
+    accumulator: fn(TResult, T) -> TResult,
+    offsets: BTreeMap<Partition, u64>,
+    batch_start_time: SystemTime,
+    message_count: usize,
+    is_complete: bool,
+}
+
+impl<T: Clone, TResult: Clone> BatchState<T, TResult> {
+    fn new(initial_value: TResult, accumulator: fn(TResult, T) -> TResult) -> BatchState<T, TResult> {
+        BatchState {
+            value: Some(initial_value),
+            accumulator,
+            offsets: Default::default(),
+            batch_start_time: SystemTime::now(),
+            message_count: 0,
+            is_complete: false,
+        }
+    }
+
+    fn add(&mut self, message: Message<T>) {
+        let mut tmp = None;
+        mem::swap(&mut tmp, &mut self.value);
+        self.value = Some((self.accumulator)(tmp.unwrap(), message.payload()));
```

Comment:
Not a fan of this but didn't find a better solution, should I switch to using `take_mut` crate or something else?

Reply:
I think this is fine. Since you're now using Option, you can actually use `self.accumulator.take()` instead of `mem::swap`